diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/NewlinesRuleWarnTest.kt

```diff
@@ -647,8 +647,154 @@ class NewlinesRuleWarnTest : LintTestBase(::NewlinesRule) {
+                |}
+            """.trimMargin(),
+                LintError(4, 10, ruleId, "$functionalStyleWarn .", true),
+                LintError(9, 10, ruleId, "$functionalStyleWarn .", true),
+                LintError(9, 16, ruleId, "$functionalStyleWarn .", true)
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.WRONG_NEWLINES)
+    fun `should warn dot qualified with save access`() {
+        lintMethod(
+                """
+                |fun foo() {
+                |   x
+                |   ?.map()
+                |   .gre().few()
+                |}
+                |
+                |fun foo() {
+                |   x
+                |   ?.map().gre().few()
+                |}
+                |fun foo() {
+                |   x
+                |   ?.map()
+                |   .gre()
+                |   .few()
+                |}
+                |
+                |fun foo() {
+                |   x?.map()
+                |   .gre()
+                |   .few()
+                |}
+            """.trimMargin(),
+                LintError(4, 10, ruleId, "$functionalStyleWarn .", true),
+                LintError(9, 11, ruleId, "$functionalStyleWarn .", true),
+                LintError(9, 17, ruleId, "$functionalStyleWarn .", true),
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.WRONG_NEWLINES)
+    fun `should warn dot qualified with exclexcl`() {
+        lintMethod(
+                """
+                |fun foo() {
+                |   x!!.map()
+                |   .gre()
+                |   .few()
+                |}
+                |
+                |fun foo() {
+                |   x!!
+                |   .map()
+                |   .gre()
+                |   .few()
+                |}
+                |
+                |fun foo() {
+                |   x!!
+                |   .map().gre()
+                |   .few()
+                |}
+                |
+                |fun foo() {
+                |   x!!.map()
+                |   .gre().few()
+                |}
+            """.trimMargin(),
+                LintError(2, 7, ruleId, "$functionalStyleWarn .", true),
+                LintError(16, 10, ruleId, "$functionalStyleWarn .", true),
+                LintError(21, 7, ruleId, "$functionalStyleWarn .", true),
+                LintError(22, 10, ruleId, "$functionalStyleWarn .", true)
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.WRONG_NEWLINES)
+    fun `should warn dot qualifided with exclexcl`() {
+        lintMethod(
+                """
+                |fun foo() {
+                |
+                |   z.goo()
+                |       ?:
+                |        goo()
+                |      
+                |   x.goo()
+                |       ?:goo()
+                |       
+                |   y.ds()?:gh()
+                |}
+            """.trimMargin(),
+                LintError(4,8, ruleId, "${WRONG_NEWLINES.warnText()} shouldn't be a new line after ?:", true)
```

Comment:
We probably need more complex logic on `?:`, like described [in the issue](https://github.com/cqfn/diKTat/issues/554#issuecomment-734722405).  I think it can be done later in a separate PR.

Reply:
Yes, we have another issue for this case